src/azure-cli-core/azure/cli/core/_profile.py

```diff
@@ -105,6 +105,25 @@ def __init__(self, storage=None, auth_ctx_factory=None, use_global_creds_cache=T
         self._management_resource_uri = CLOUD.endpoints.management
         self._ad_resource_uri = CLOUD.endpoints.active_directory_resource_id
 
+    def refresh_subscriptions(self):
+        # Currently requires a default subscription selected
+        existingSubscriptions = self.load_cached_subscriptions()
+        tenants = {}
+        bearer = self.get_raw_token()
```

Comment:
Under this flow, we should not manipulate tokens directly. Only `az login` uses naked token because there might have no established creds, but for `refresh`, we do have. 
Here is the what I would do in a normal and simple way. (The `TODOs` below skip a few trivial logics to get my main idea clear)
```python
        from azure.mgmt.resource import SubscriptionClient

        subscriptions = self.load_cached_subscriptions()
        client_infos = []
        # TODO: eliminate redundant clients from the same tenant
        for s in subscriptions:
            client_infos.append((SubscriptionClient(self.get_login_credentials()[0]),s))

        refreshed_subs = []
        for c in client_infos:
            result = list(c[0].subscriptions.list())
            for r in result:
                setattr(r, 'tenant_id', c[1][_TENANT_ID])
            refreshed_subs += Profile._normalize_properties(c[1][_USER_ENTITY][_USER_NAME],
                                                            result,
                                                            c[1][_USER_ENTITY][_USER_TYPE]==_SERVICE_PRINCIPAL)

        # TODO: 
        # a: extract out following common code with "find_subscriptions_on_login", 
        # b. handle 'tenant only accounts' (no subscriptions)
        # c: ensure we maintain the active subscription
        self._set_subscriptions(refreshed_subs)
        return deepcopy(refreshed_subs)
```